### PR TITLE
refactor: Revamp the project

### PR DESCRIPTION
**Pull request by damianopetrungaro**

### CODE REVIEW

**Comment by damianopetrungaro** [comment]
@ElenaForester @mereba please review this ***_HUGE_ PR***.
Please if you are free to help me to increase the coverage at 100% (and take a look if the current tests can be revamped too, they are quite old).

💪

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,39 @@
+FROM composer:latest as composer
+FROM php:fpm
+
+# Define ENVs and ARGs
+ARG XDEBUG_REMOTE_HOST=docker.for.mac.localhost
```
Comment (mereba) [inline Dockerfile:5]: Maybe we can remove this env var and others as well from the Dockerfile and allow non-mac devs to specify it in their docker-compose.yml?

File: src/Adapters/DbalAdapter.php [hunk 2]
```diff
@@ -35,42 +34,36 @@ class DbalAdapter extends AbstractAdapter
...
+    public function __construct(
+        string $poolName,
+        Connection $connection,
+        array $tableOptions = [],
+        $expirationDateFormat = null
+    ) {
         $this->poolName = $poolName;
         $this->connection = $connection;
-        $this->tableOptions = array_merge($this->tableOptions, $tableOptions);
+        $this->tableOptions = \array_merge($this->tableOptions, $tableOptions);
```
Comment (mereba) [inline src/Adapters/DbalAdapter.php:55]: Why are you backslashing all the global functions and classes? Have you redefined them in your namespace?

File: src/Adapters/RedisAdapter.php [hunk 3]
```diff
@@ -20,156 +20,131 @@ class RedisAdapter extends AbstractAdapter
...
+        return (bool) $this->redis->delete($key);
     }
 
-    /**
-     * {@inheritdoc}
-     */
     public function deleteItems(array $keys): bool
     {
-        $this->normalizeKeyName($keys);
+        $keys = $this->normalizeKeyNames($keys);
```
Comment (mereba) [inline src/Adapters/RedisAdapter.php]: Probably rename $keys here? `$keys` as a parameter to `normalizeKeyNames()` is different from `$keys` as a newly created variable to which the result of `normalizeKeyNames()` is assigned. So maybe better to explicitly state this? With something like `$normalizedKeys = $this->normalizeKeyNames($keys);`. Easier to read and could potentially avoid side effects.

File: src/Adapters/RedisAdapter.php [hunk 4]
```diff
@@ -20,156 +20,131 @@ class RedisAdapter extends AbstractAdapter
...
+
+        return (bool) $this->redis->delete($keys);
     }
 
-    /**
-     * {@inheritdoc}
-     */
     public function deleteItem(string $key): bool
     {
         $this->normalizeKeyName($key);
```
Comment (mereba) [inline src/Adapters/RedisAdapter.php]: Did you forget to assign this to a new variable?

File: src/CacheItem.php [hunk 5]
```diff
@@ -4,116 +4,99 @@
...
         return $this;
     }
 
-    /**
-     * {@inheritdoc}
-     */
     public function expiresAt($expiration = null): CacheItemInterface
     {
-        if ($expiration instanceof \DateTimeInterface) {
+        if ($expiration instanceof \DateTimeImmutable) {
```
Comment (mereba) [inline src/CacheItem.php:71]: Using switch ... case here might look easier on the eyes ;-)

File: src/CacheItemPool.php [hunk 6]
```diff
@@ -17,145 +16,107 @@ class CacheItemPool implements CacheItemPoolInterface
...
      */
-    protected $adapterInterface;
+    protected $adapter;
 
     /**
-     * List of items waiting to be saved
+     * List of items waiting to be saved.
      *
-     * @var CacheItemCollectionInterface $deferredItems
+     * <<<<<<< HEAD
```
Comment (mereba) [inline src/CacheItemPool.php]: Forgot to clean up after fixing merge conflict?

File: src/Adapters/RedisAdapter.php [hunk 7]
```diff
@@ -20,156 +20,131 @@ class RedisAdapter extends AbstractAdapter
...
-        } else {
-            $keys = $this->poolName . $this->separator . $keys;
-        }
+        return \array_map([$this, 'normalizeKeyName'], $keys);
+    }
+
+    /**
+     * Normalize the key, adding the poolName prefix.
+     */
+    protected function normalizeKeyName(string $keys): string
```
Comment (mereba) [inline src/Adapters/RedisAdapter.php]: normaliseKeyName($key~s~)

File: .gitignore [hunk 8]
```diff
@@ -1,4 +1,2 @@
-composer.phar
-composer.lock
```
Comment (mereba) [inline .gitignore:2]: Nice!

**Review by mereba — COMMENTED** [review]
Great job man! Indeed, huuuuge PR. Maybe submit smaller PRs in the future? :-p

File: Dockerfile (hunk 1 above)
Comment (damianopetrungaro) [reply Dockerfile:5]: It's possible to override the value using the `args` in the build phase 😄 

https://docs.docker.com/compose/compose-file/#args
So also non-mac user can use the `Dockerfile`/`docker-compose.yml`

File: src/Adapters/DbalAdapter.php (hunk 2 above)
Comment (damianopetrungaro) [reply src/Adapters/DbalAdapter.php:55]: Using the backslash we enforce the usage of the built-in PHP functions, avoiding possible overrides by userland.

There's also a good performance improvement about using the fully qualified function calls.

File: src/Adapters/RedisAdapter.php (hunk 4 above)
Comment (damianopetrungaro) [reply src/Adapters/RedisAdapter.php]: Good catch!

File: src/Adapters/RedisAdapter.php (hunk 3 above)
Comment (damianopetrungaro) [reply src/Adapters/RedisAdapter.php]: Partially agree, there's a micro-optimization using the same variable name (overriding the memory address will cost less memory) but it won't be a huge impact!

File: src/CacheItem.php (hunk 5 above)
Comment (damianopetrungaro) [reply src/CacheItem.php:71]: I don't see any benefit about this 😄 
This is just a personal preference 😛

**Comment by damianopetrungaro** [comment]
@mereba  
> Great job man! Indeed, huuuuge PR. Maybe submit smaller PRs in the future? :-p

Yeah, for sure, this is a PR with a rebase of more than 7 commits never merged in the past, as you can see!

**Pull request merged**